sigs/jvm/rfcs/20190904-java-tensors.md

```diff
@@ -0,0 +1,301 @@
+
+The underlying implementation though must be changed to avoid using reflective techniques to generate the tensor and
+must be more explicit. To simplify this task, we should limit the possible factories to rank-0 or rank-1
+constants (which satisfies most of the cases) and another to pass explicitly a `Tensor` instance for constant of a 
+higher rank. For example:
+
+```java
+public final class Constant<U> extends PrimitiveOp {
+
+    public static final Constant<TInt32> create(Scope scope, Integer value) {
+        ...
+    }
+    
+    public static final Constant<TInt32> create(Scope scope, Integer... values) {
+        ...
+    }
+    
+    public static final <U extends Tensor<?>> Constant<U> create(Scope scope, U data) {
+        ...
+    }
+}
+
+Constant<TInt32> scalar = tf.constant(10);  // single argument constructor is chosen
+
+Constant<TInt32> vector = tf.constant(10, 11, 42);  // variadic constructor is chosen
+
+Constant<TInt32> matrix = tf.constant(TInt32.ofShape(2, 2).row(10, 11).row(30, 40).done());
+
+```
+
+## Addendum
+
+### NdArray Iterative Data Initialization
+
+In some previous examples, some tensors are initialize using a new method called `row`. This method is intented
+to be added to the original [`NdArray` API](https://github.com/karllessard/community/blob/master/sigs/jvm/rfcs/20190606-java-tensor-io.md#ndarray-api)
+submitted in the previous RFC, with the objective of allowing the allocation and initialization
+of a small tensor in a single line to mimics standard multidimensional Java arrays.
+
+The idea is to initialize the data of each row of the `NdArray` sequentially by invoking `row(...)`, 
+and then invoke the `done()` method to return the newly allocated `NdArray` instance. A "row" in this terminology
+means a vector on the last axis (dimension) of an array.
+
+Let's compare standard Java arrays with those new methods as they are used with tensors:
+```java
+
+// Create a 2D matrix
+
+int[][] matrix2d_s = new int[][] { {1, 2}, {3, 4} };
+
+TInt32 matrix2d_t = TInt32.ofShape(2, 2).row(1, 2).row(3, 4).done();
+
+// Create a 3D matrix
+
+int[][][] matrix3d_s = new int[][][] { { {1, 2}, {3, 4} }, { {5, 6}, {7, 8} }, { {9, 10}, {11, 12} } };
+
+TInt32 matrix3d_t = TInt32.ofShape(3, 2, 2)
+  .row(1, 2).row(3, 4)
+  .row(5, 6).row(7, 8)
+  .row(9, 10).row(11, 12)
+  .done();
+```
+Of course, the standard Java arrays gives a better view on where is located a given row, by embedding it
+through a series of brackets. The actual proposal is focusing on simplicity and performance more that readability.
+*Please, feel free to suggest any better solution.*
+
+Note also that an advantage of using `NdArray` is that the type and the shape of the tensor is explicit and can be easily 
+retrieved by calling `t.dataType()` and `t.shape()` respectively, while with standard Java arrays those need
+to be inferred and discovered by the TF Java client using costly reflections and navigating through the array.
```

Comment:
It's not a reflective operation that takes time, it's just the physical traversal of the Java multidimensional array, as they could be stored anywhere on the heap, and a 4d array could be made up of hundreds of thousands of arrays.

Reply:
ok I'll change the text to avoid the confusion.